### DISCUSSION TITLE
Path is not represented in current version locations

### DISCUSSION THREAD

**Discussion by harpener** [sqlalchemy/alembic#1284] [Usage Questions]
Hi guys.

**Describe the bug**

Cannot generate migration in project with multiple heads because location paths do not match.

**Expected behavior**

Migration is generated.

**To Reproduce**

```
alembic revision --autogenerate --head "after@head" -m "my_migration"
alembic revision --autogenerate --head "before@head" -m "my_migration"
```

**Error**

```
FAILED: Path C:\...\alembic\after is not represented in current version locations
FAILED: Path C:\...\alembic\before is not represented in current version locations
```

**Versions.**
 - OS: Windows 11 Pro, 22H2, 22621.1992
 - Python: 3.8.17
 - Alembic: 1.11.1
 - SQLAlchemy: 1.4.48
 - Database: PostgreSQL 9.6.2

**Additional context**

I use two heads - `after` and `before`. I have following in `alembic.ini`:
```
version_locations = %(here)s/alembic/before %(here)s/alembic/after
```

I noticed that my Python returns following:
```
>>> import os.path

>>> os.path.abspath("")
'c:\\...'

>>> os.path.realpath("")
'C:\\...'
```

As you can see, the letter-case of drive letter is different which is where the problem comes from.

I haven't encountered such error in past, but I recently installed the OS anew and updated Python to its latest security fix, if it helps.

Do you have any idea what is wrong?

**Have a nice day!**

**Comment by CaselIT** [comment]
Hi,

What are using as the setting of version_path_separator?

**Reply by harpener** [reply]
Hi, nothing, it isn't specified in INI file, it should be a default I guess.

**Reply by CaselIT** [reply]
The default should be os then, try using space if you are using it as separator

**Reply by harpener** [reply]
It does not help. Anyway, according to the documentation, the default should handle spaces: https://github.com/sqlalchemy/alembic/blob/rel_1_11_1/alembic/templates/multidb/alembic.ini.mako#L46

**Reply by CaselIT** [reply]
Strange, does the issue happens if you use a relative path instead of using `here`?

**Reply by harpener** [reply]
The issue persists also for relative path like `version_locations = alembic/before alembic/after`

**Reply by CaselIT** [reply]
That's strange, can you provide a complete example with the env.py, the alembic.ini and some sample migration that illustrate the issue?

**Reply by harpener** [reply]
Hi, so I found out where the problem originates finally.

I am using `virtualenvwrapper-win` to manage virtual environments and for each project I use `setprojectdir` to store its workspace location, so that it is loaded everytime I switch into that virtual environment. Some of my projects had small drive letter in the workspace path, which is something I noticed in console prompt. As I pointed out in the issue description, command `os.path.abspath` returns the same drive letter casing as currently selected, whereas `os.path.realpath` always returns upper-case drive letter. Matching the locations case-sensitively, together with the fact that head parameter is translated into path using `os.path.realpath` [here](https://github.com/sqlalchemy/alembic/blob/rel_1_11_1/alembic/script/base.py#L139) and version locations are translated into path using `%(here)s` which is using only `os.path.abspath` [here](https://github.com/sqlalchemy/alembic/blob/rel_1_11_1/alembic/config.py#L198), is what was causing an issue on my side. I fixed the workspace location, but I recommend fixing this also in Alembic, if you want to be correct.

Thanks for the help, feel free to close the discussion.

**Reply by CaselIT** [reply]
What may be missing is a call to normcase here https://github.com/sqlalchemy/alembic/blob/0064c5183cc19ee84954f75f44b4b5f9131973fb/alembic/script/base.py#L139

can you check if chaning that line to `real_path = os.path.normcase(os.path.realpath(file_path))` fixes this issue?

if so we can then came with a test for this

**Reply by harpener** [reply]
It does not solve the issue. `os.path.normcase` makes all characters lower-case, not just the drive letter. The version location is matched case-sensitively.

**Reply by zzzeek** [reply]
hey there -

by what means is this a Windows file system where paths are case sensitive?  that's not windows behavior at all.

**Reply by zzzeek** [reply]
also if windows file paths are case sensitive with python, then this is a bug in python:

```
>>> import os.path

>>> os.path.abspath("")
'c:\\...'

>>> os.path.realpath("")
'C:\\...'
```

how would that not be a bug?

**Reply by harpener** [reply]
Hi, maybe it is Python, but I think I read that Python calls underlying system libraries, so maybe it could be just that. Anyhow, the problem is not that you cannot access given path case-insensitively, it is that you are comparing two literal strings that are case sensitive.

**Reply by zzzeek** [reply]
it makes no sense that the "c:" is changing case

**Reply by zzzeek** [reply]
anyway, patches welcome for whatever approach anyone thinks is the "correct" way to do this that passes tests and does not break case sensitive paths on other OSes

**Reply by harpener** [reply]
It's not the real path, so it can make sense in workspace context. It's the same with symlinks, I am confident that if I put a symlink to a directory in the config file, it will not match the version locations again, even in linux. IMO all that is needed is to translate the `%(here)s` using the real path, since it is compared to a real path later as well. Thanks for the input.

**Reply by CaselIT** [reply]
that may be an option. I'll take a look to see if it's a possible change

**Reply by zzzeek** [reply]
see #1330

**Comment by CaselIT** [comment]
this seems actually an issue so re-opening it as issue #1366
